File: spring/src/test/java/cardsproject/controller/marketplace/TradeDisputeControllerTest.java
package cardsproject.controller.marketplace;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.transaction.annotation.Transactional;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

@SpringBootTest
@AutoConfigureMockMvc
@Transactional
public class TradeDisputeControllerTest {

    @Autowired
    private MockMvc mockMvc;

    @Test
    void list_returns200() throws Exception {
        mockMvc.perform(get("/api/trade_disputes"))
            .andExpect(status().isOk());
    }
    @Test
    void create_returns201() throws Exception {
        mockMvc.perform(post("/api/trade_disputes")
            .contentType(MediaType.APPLICATION_JSON)
            .content("{ \"reason\": \"ITEMNOTRECEIVED\", \"description\": \"test\", \"openedAt\": \"2024-01-01T00:00:00\", \"resolvedAt\": null }"))
            .andExpect(status().isCreated());
    }
    @Test
    void show_returns200or404() throws Exception {
        mockMvc.perform(get("/api/trade_disputes/1"))
            .andExpect(result -> {
                int status = result.getResponse().getStatus();
                assert status == 200 || status == 404 || status == 403;
            });
    }
    @Test
    void create_fails_when_resolved_at_requires_terminal_status_violated() throws Exception {
        // resolved_at_requires_terminal_status: antecedent true, consequent missing → 400
        mockMvc.perform(post("/api/trade_disputes")
            .contentType(MediaType.APPLICATION_JSON)
            .content("{ \"status\": \"OPEN\", \"reason\": \"ITEMNOTRECEIVED\", \"description\": \"test\", \"openedAt\": \"2024-01-01T00:00:00\", \"transactionId\": 1, \"openedById\": 1, \"resolvedAt\": \"2024-01-01T00:00:00\" }"))
            .andExpect(status().isBadRequest());
    }
    @Test
    @org.springframework.security.test.context.support.WithMockUser(roles = {"ADMIN", "MODERATOR"})
    void transitionOpenToUnderReview_returns200or404() throws Exception {
        mockMvc.perform(patch("/api/trade_disputes/1/transitions/open-to-underreview"))
            .andExpect(result -> {
                int s = result.getResponse().getStatus();
                assert s == 200 || s == 404 || s == 409 || s == 422;
            });
    }

    @Test
    @org.springframework.security.test.context.support.WithMockUser(roles = {"ADMIN", "MODERATOR"})
    void transitionUnderReviewToResolved_returns200or404() throws Exception {
        mockMvc.perform(patch("/api/trade_disputes/1/transitions/underreview-to-resolved"))
            .andExpect(result -> {
                int s = result.getResponse().getStatus();
                assert s == 200 || s == 404 || s == 409 || s == 422;
            });
    }

    @Test
    @org.springframework.security.test.context.support.WithMockUser(roles = {"ADMIN"})
    void transitionUnderReviewToEscalated_returns200or404() throws Exception {
        mockMvc.perform(patch("/api/trade_disputes/1/transitions/underreview-to-escalated"))
            .andExpect(result -> {
                int s = result.getResponse().getStatus();
                assert s == 200 || s == 404 || s == 409 || s == 422;
            });
    }

    @Test
    @org.springframework.security.test.context.support.WithMockUser(roles = {"ADMIN"})
    void transitionEscalatedToResolved_returns200or404() throws Exception {
        mockMvc.perform(patch("/api/trade_disputes/1/transitions/escalated-to-resolved"))
            .andExpect(result -> {
                int s = result.getResponse().getStatus();
                assert s == 200 || s == 404 || s == 409 || s == 422;
            });
    }

    @Test
    void transitionResolvedToOpen_isDenied() throws Exception {
        mockMvc.perform(patch("/api/trade_disputes/1/transitions/resolved-to-open"))
            .andExpect(result -> {
                int s = result.getResponse().getStatus();
                assert s == 409 || s == 404;
            });
    }
}

File: phoenix/test/marketplace/trade_dispute_controller_test.exs
defmodule CardsProjectWeb.Marketplace.TradeDisputeControllerTest do
  use ExUnit.Case, async: false
  use CardsProjectWeb.ConnCase

  @valid_params %{
    "description" => "test",
    "opened_at" => ~N[2024-01-01 00:00:00],
    "status" => "Open",
    "reason" => "ItemNotReceived"
  }

  describe "GET /api/trade_disputes" do
    test "returns 200 with list", %{conn: conn} do
      conn = get(conn, "/api/trade_disputes")
      assert json_response(conn, 200) |> is_list()
    end
  end

  describe "POST /api/trade_disputes" do
    test "creates record and returns 201", %{conn: conn} do
      conn = post(conn, "/api/trade_disputes", @valid_params)
      assert %{"id" => _id} = json_response(conn, 201)
    end
  end

  describe "GET /api/trade_disputes/:id" do
    test "returns 200 for existing record", %{conn: conn} do
      {:ok, record} = CardsProject.Marketplace.create_trade_dispute(@valid_params)
      conn = get(conn, "/api/trade_disputes/#{record.id}")
      assert json_response(conn, 200)
    end
  end

  describe "PUT /api/trade_disputes/:id" do
    test "updates and returns 200", %{conn: conn} do
      {:ok, record} = CardsProject.Marketplace.create_trade_dispute(@valid_params)
      conn = put(conn, "/api/trade_disputes/#{record.id}", @valid_params)
      assert json_response(conn, 200)
    end
  end

  describe "DELETE /api/trade_disputes/:id" do
    test "deletes and returns 204", %{conn: conn} do
      {:ok, record} = CardsProject.Marketplace.create_trade_dispute(@valid_params)
      conn = delete(conn, "/api/trade_disputes/#{record.id}")
      assert response(conn, 204)
    end
  end

  describe "PATCH /api/trade_disputes/:id/transitions/open-to-underreview" do
    test "transitions Open -> UnderReview", %{conn: conn} do
      {:ok, record} = CardsProject.Marketplace.create_trade_dispute(@valid_params)
      conn = patch(conn, "/api/trade_disputes/#{record.id}/transitions/open-to-underreview")
      assert conn.status in [200, 409, 422, 404]
    end
  end

  describe "PATCH /api/trade_disputes/:id/transitions/underreview-to-resolved" do
    test "transitions UnderReview -> Resolved", %{conn: conn} do
      {:ok, record} = CardsProject.Marketplace.create_trade_dispute(@valid_params)
      conn = patch(conn, "/api/trade_disputes/#{record.id}/transitions/underreview-to-resolved")
      assert conn.status in [200, 409, 422, 404]
    end
  end

  describe "PATCH /api/trade_disputes/:id/transitions/underreview-to-escalated" do
    test "transitions UnderReview -> Escalated", %{conn: conn} do
      {:ok, record} = CardsProject.Marketplace.create_trade_dispute(@valid_params)
      conn = patch(conn, "/api/trade_disputes/#{record.id}/transitions/underreview-to-escalated")
      assert conn.status in [200, 409, 422, 404]
    end
  end

  describe "PATCH /api/trade_disputes/:id/transitions/escalated-to-resolved" do
    test "transitions Escalated -> Resolved", %{conn: conn} do
      {:ok, record} = CardsProject.Marketplace.create_trade_dispute(@valid_params)
      conn = patch(conn, "/api/trade_disputes/#{record.id}/transitions/escalated-to-resolved")
      assert conn.status in [200, 409, 422, 404]
    end
  end

  describe "PATCH /api/trade_disputes/:id/transitions/resolved-to-open" do
    test "is denied with 409", %{conn: conn} do
      {:ok, record} = CardsProject.Marketplace.create_trade_dispute(@valid_params)
      conn = patch(conn, "/api/trade_disputes/#{record.id}/transitions/resolved-to-open")
      assert conn.status in [409, 404]
    end
  end

end
